### ISSUE TITLE
Intel STM/PPAM support

### ISSUE THREAD

**Issue by miczyg1** [Dasharo/dasharo-issues#132]
**The problem you're addressing (if any)**
Currently Dasharo firmware does not support Intel PPAM.

**Describe the solution you'd like**
Implement PPAM support.

**Where is the value to a user, and who might that user be?**
Better security,  DRTM compatibility for Microsoft Secured Core PC.

**Describe alternatives you've considered**
Integrate STM instead of PPAM

**Additional context**
None

**Comment by bdelgado1995** [comment]
coreboot supports the STM which is a useful mechanism to run SMI handlers in a virtual machine. Enabling this feature could help protect against runtime SMM vulnerabilities.

**Comment by pietrushnic** [comment]
@bdelgado1995 yes we know about STM and even met creator of the code couple times. He was interested about our work related to [TrenchBoot](https://trenchboot.org/) project. AFAIK STM is hardware-specific feature and was tested only on Skylake. I guess it is not just enable that features and it works since it is highly hardware dependent. @miczyg1 please correct me if I'm wrong. Also STM is probably way more complex concept then PPAM IIUC.

**Comment by miczyg1** [comment]
> I guess it is not just enable that features and it works since it is highly hardware dependent

It is not that much hardware dependent... Although some glue would be needed to support different microarchitectures. STM is also typically paired with TXT.

Secondly, any OS you would like to run on STM-enabled machine needs to know that STM is out there.
Thirdly,  any OS you would like to run on STM-enabled must have some interface to STM implemented.

Because of those two reasons STM did not meet wide adoption. Also STM introduces a performance penalty.

STM/PPAM makes very much sense with proprietary BIOSes, where SMM is an unknown black hole. With coreboot we have fully open and auditable SMM code where STM would not add that much more security.

**Comment by bdelgado1995** [comment]
Good points. 

There are some community-written STM launchers for [Xen ](https://github.com/PPerfLab/xen/tree/stm-optin)and [Linux](https://github.com/EugeneDMyers/stm_linux_module). 

For performance, there is some [performance impact](http://web.cecs.pdx.edu/~karavan/research/HASP20_preprint_Delgado.pdf) when there is an SMI. Some systems can go a long period of time without SMIs in which case there is no perf impact and others have some a few times a second so each SMI would take a little longer. Linking to a paper our team wrote that gives some coverage of this. 

I agree it makes sense to consider the particular SMI handlers in use and make a determination about ROI.

**Comment by miczyg1** [comment]
> Good points.
> 
> There are some community-written STM launchers for [Xen ](https://github.com/PPerfLab/xen/tree/stm-optin)and [Linux](https://github.com/EugeneDMyers/stm_linux_module).

Thank you for these pointers, these will definitely help with STM testing.

**Comment by miczyg1** [comment]
Added STM as an alternative to the description

**Comment by bdelgado1995** [comment]
If you would like to talk with an Intel business rep on the Intel System Resources Defense/PPAM, I can do some checking and see who could provide more info. Happy to try to get more info so you can get a sense of all of the options.

**Comment by pietrushnic** [comment]
@bdelgado1995, could you please set up a call for TrenchBoot Committee (Daniel P. Smith - Apertus Solutions, Rich Persaud - OpenXT, Ross Philipson - Oracle, and me)?

**Comment by bdelgado1995** [comment]
Yes. I'll do some checking. May take a little time to find the right contact, will update when I have info.

**Comment by bdelgado1995** [comment]
[Building - Z690A - STM - 2024.md](https://github.com/Dasharo/dasharo-issues/files/15154910/Building.-.Z690A.-.STM.-.2024.md)
[savedStmConfig.txt](https://github.com/Dasharo/dasharo-issues/files/15154930/savedStmConfig.txt)

I have put together an experimental setup guide for the STM build/load/launch in the Markdown file. I have also attached the .config file used (called savedStmConfig.txt). 

It would be excellent if the Dasharo devs could try this out and also streamline the installation. 
For everyone else, please use at your own risk, this is experimental and you may need to recover your system with MSI FlashBack/hardware reflasher if things go awry. 

If there are tweaks/suggestions/questions for the guide, feel free to let me know and I can update it accordingly.

**Comment by pietrushnic** [comment]
@BeataZdunczyk @macpijan, please let me know what budget will be needed here to consider this feature.

**Comment by pietrushnic** [comment]
@BeataZdunczyk @macpijan I would appreciate your attention here. @filipleple, maybe it makes sense to add this for Dell OptiPlex's upcoming release?

**Comment by miczyg1** [comment]
> [Building - Z690A - STM - 2024.md](https://github.com/Dasharo/dasharo-issues/files/15154910/Building.-.Z690A.-.STM.-.2024.md) [savedStmConfig.txt](https://github.com/Dasharo/dasharo-issues/files/15154930/savedStmConfig.txt)
> 
> I have put together an experimental setup guide for the STM build/load/launch in the Markdown file. I have also attached the .config file used (called savedStmConfig.txt).
> 
> It would be excellent if the Dasharo devs could try this out and also streamline the installation. For everyone else, please use at your own risk, this is experimental and you may need to recover your system with MSI FlashBack/hardware reflasher if things go awry.
> 
> If there are tweaks/suggestions/questions for the guide, feel free to let me know and I can update it accordingly.

I'm looking into integrating the STM in Dasharo firmware by default on the MSI desktops, as mentioned here: https://github.com/Dasharo/coreboot/pull/431#issuecomment-4646378520

@bdelgado1995  I have some questions to your guide. 

1. Does the STM necessarily have to use a PCI extension card for debug? 

We have a legacy 8250/16550 UART at I/O port 0x3f8, which should also be suitable for debugging. There is a UART TX pin exposed on the board (JBD1 header): https://docs.dasharo.com/unified/msi/development/#serial-debug

2. I see a lot of `STM Page Fault`s in the log. Is this something to fix in coreboot? Missing resources not reported to STM?

**Comment by bdelgado1995** [comment]
Hi @miczyg1 . For #1, it doesn't have to be a PCI serial card if you have another workable one. For #2, The STM for coreboot maps the first 4 GB of RAM for itself but page faults to bring in additional pages that are outside of this range as-needed. (This mapping behavior is one difference between the original STM for UEFI (https://github.com/jyao1/STM) and the version for coreboot).
